File: 06-week-5--big-o-continued/00-days-1-to-2--implement-a-stack-class/ruby/stack.rb
class Stack
  attr_reader :limit

  def initialize
    @stack = []
    # this is an arbitrary value to make testing easier
    # 1,024 would be way too high to test!
    @limit = 10
  end

  # add item to top of stack if not full
  # if full, throw error
  def push(item)
  end

  # remove item from top of stack and return it
  def pop
  end

  # return item at top of stack without removing it
  def peek
  end

  # return true if stack is empty, otherwise false
  def isEmpty?
  end

  # return true if stack is full, otherwise false
  def isFull?
  end

  # return number of items in stack
  def size
  end

  # return -1 if item not in stack, otherwise integer representing 
  # how far it is from the top
  def search(target)
  end

  # print contents of stack: do not return the stack itself!
  def print
  end
end

if __FILE__ == $PROGRAM_NAME
  # Don't forget to add your tests!
end

File: 06-week-5--big-o-continued/00-days-1-to-2--implement-a-stack-class/javascript/stack.js
class Stack {
  constructor() {
    this.stack = [];
    // this is an arbitrary value to make testing easier
    // 1,024 would be way too high to test!
    this.limit = 10;
  }

  // add item to top of stack if not full
  // if full throw error
  push(item) {
    if (!this.isFull()) {
      this.stack.push(item)
    } else {
      throw new Error("Stack is full!")
    }
  }

  // remove item from top of stack and return it
  pop() {
    return this.stack.pop()
  }

  // return item at top of stack without removing it
  peek() {
    return this.stack[this.size() -1]
  }

  // return true if stack is empty, otherwise false
  isEmpty() {
   return this.size() === 0
  }

  // return true if stack is full, otherwise false
  isFull() {
    return this.size() === this.limit
  }

  // return number of items in stack
  size() {
    return this.stack.length
  }

  // return -1 if item not in stack, otherwise integer representing 
  // how far it is from the top
  search(target) {
    for (let i = -1; i >= -this.size(); --i) {
      if (this.stack[this.size() + i] === target) {
        return Math.abs(i) - 1;
      }
    }

    return -1;
  }


  // print contents of stack: do not return the stack itself!
  print() {
    console.log(this.stack.join(' <- '))
  }
}


if (require.main === module) {
  // add your own tests in here
}

module.exports = Stack;
